### PR TITLE
Fix Gradient Accumulation issue

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
First draft

End goal is to make it easy for anyone to:
- change the loss for his model 
- contribute a new loss for a model (like vision model, ENCODEC etc)
- allow passing arbitrary kwargs, interfacing 

TODO:
- [ ] Fix deformable detr loss computation

### CODE REVIEW

File: src/transformers/loss_utils.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    num_items = kwargs.pop("num_items", None)
+
+    if num_items is not None:
+        # Calculate the CrossEntropyLoss manually when using grad accum
+        log_probs = nn.functional.log_softmax(shift_logits, dim=-1)
+        loss = -log_probs[range(shift_labels.size(0)), shift_labels]
+        loss = loss.sum() / num_items
+    else:
+        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100)
```
Comment (kashif) [inline src/transformers/loss_utils.py:29]:
```suggestion
    if num_items is not None:
        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
        loss = loss / num_items
    else:
        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100)
```

File: src/transformers/loss_utils.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    logits = logits.view(-1, config.num_labels)
+    labels = labels.view(-1)
+    logits = logits.float()
+    # Flatten the tokens
+    loss_fct = CrossEntropyLoss()
+    return loss_fct(logits, labels)
+
+
+LOSS_MAPPING = {
+    "ForCausalLM": DefaultCrossEntropyLoss,
```
Comment (BenjaminBossan) [inline src/transformers/loss_utils.py:91]: Just wondering aloud: Instead of matching based on class name, could we do a mapping from class to loss, and then do something like:

```python
for key, val in LOSS_MAPPING.items():
    if isinstance(self, key):
        loss = val
        break
else:  # no break
    # raise error
```

I assume the matching exists for custom classes that are out there in the wild. If name is a more reliable predictor than inheritance or if I'm misunderstanding, please disregard my comment.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for coming forward with this fix so quickly. There is probably not much I can help with, but I took a look and added some comments.

File: src/transformers/loss_utils.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/loss_utils.py:91]: Will look into improving, but this looks super slow

Comment (BenjaminBossan) [reply src/transformers/loss_utils.py:91]: This would only be run once because of the LRU cache, right?

File: src/transformers/loss_utils.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/loss_utils.py:30]: @muellerzr as you wrote the snippet!

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -4979,6 +4980,32 @@ def _is_quantized_training_enabled(self):
...
+                loss_groups = f"({'|'.join(LOSS_MAPPING)})"
+                loss_type = re.findall(loss_groups, self.__class__.__name__)
+                if len(loss_type) > 0:
+                    loss_type = loss_type[0]
+                else:
+                    loss_type = None
+        if loss_type is None:
+            raise ValueError(
+                "We could not determine which loss function to use."
+                f"based on the the class name. Make sure you add `{ self.__class__.__name__}` to the `LOSS_MAPPING`"
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/modeling_utils.py:5000]: Yep will update this

File: src/transformers/loss_utils.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/loss_utils.py]: Also we get the classname from the class itself and want to have good defaults instead of matching against full name!

File: src/transformers/loss_utils.py (hunk 1 above)
Comment (kashif) [reply src/transformers/loss_utils.py]: i believe currently it is not ignoring the losses and potentially the functional API might be faster?
```
loss.masked_fill_(shift_labels == -100, 0)
```

File: src/transformers/loss/loss_utils.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    # Enable model parallelism
+    shift_labels = shift_labels.to(shift_logits.device)
+
+    num_items = kwargs.pop("num_items", None)
+
+    if num_items is not None:
+        # Calculate the CrossEntropyLoss manually when using grad accum
+        log_probs = nn.functional.log_softmax(shift_logits, dim=-1)
+        loss = -log_probs[range(shift_labels.size(0)), shift_labels]
+        loss = loss.sum() / num_items
```
Comment (kashif) [inline src/transformers/loss/loss_utils.py:41]:
```suggestion
    if num_items is not None:
        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
        loss = loss / num_items
```

File: src/transformers/loss_utils.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/loss_utils.py]: aligned with the suggestion here!

**Review by LysandreJik — APPROVED** [review]
Very good! I'll ask Daniel if he's down to review, it would be very useful to have his opinion. Thanks

File: src/transformers/loss/loss_rt_detr.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+
+from ..utils import is_accelerate_available, is_scipy_available, is_vision_available, requires_backends
+from .loss_for_object_detection import (
+    _set_aux_loss,
+    generalized_box_iou,
+    linear_sum_assignment,
+)
+from .loss_rt_detr import DeformableDetrImageLoss
```
Comment (kashif) [inline src/transformers/loss/loss_rt_detr.py:25]:
```suggestion
from .loss_deformable_detr import DeformableDetrImageLoss
from .loss_for_object_detection import (
    _set_aux_loss,
    generalized_box_iou,
    linear_sum_assignment,
)
```

**Comment by danielhanchen** [comment]
Continuing discussion from https://github.com/huggingface/transformers/pull/34198 - @muellerzr is correct if truncate and partition all sequences to block sizes of 128 for Wikitext, the GA fix and without the GA fix should nearly be equivalent. I can also repro this:
image

But unfortunately if we don't pad it and use varying sequence lengths, the loss for `bsz = 1, ga = 32` actually seems to go haywire:
image

If you plot the training loss for `bsz = 1, ga = 32` with and without the fix (blue and yellow), and also `bsz = 32, ga = 1` (red) under a log scale, we get:
image

And if we plot them on the left and right axes:
image

All experiments here: https://docs.google.com/spreadsheets/d/1RUiVuFNfnl9eBAa3JhvkKb0hm20m4NqnUO-OWDPpNos/edit?usp=sharing - also the `bsz = 32` varying sequence lengths unfortunately OOMed on a A100 40GB, so only 50% of the training went through.

**Comment by muellerzr** [comment]
Can also confirm, and confirm the fix in #34198 solves it
W B Chart 10_17_2024, 5_45_37 AM](https://github.com/user-attachments/assets/c21dc2fb-2e11-4c3b-917d-a6e4cc5bfd9b)
f9726-dfb3-4efb-aa61-0a4962d02f39)

File: src/transformers/loss/loss_utils.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    shift_logits = shift_logits.view(-1, vocab_size)
+    shift_labels = shift_labels.view(-1)
+    # Enable model parallelism
+    shift_labels = shift_labels.to(shift_logits.device)
+
+    num_items = kwargs.pop("num_items", None)
+
+    if num_items is not None:
+        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
+        loss = loss / num_items
```
Comment (kashif) [inline src/transformers/loss/loss_utils.py:41]:
```suggestion
    num_items_in_batch = kwargs.pop("num_items_in_batch", None)

    if num_items_in_batch is not None:
        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
        loss = loss / num_items_in_batch
```

Comment (aurelienmorgan) [reply src/transformers/loss/loss_utils.py:41]: ain't there a missing space in
```python
if num_items_in_total_batchis not None:
```
?

Comment (kashif) [reply src/transformers/loss/loss_utils.py:41]: thanks! also do we want to check if `num_items_in_total_batch` is not zero?

File: src/transformers/loss/loss_utils.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+    num_items_in_batch = kwargs.pop("num_items_in_batch", None)
+
+    if num_items_in_batch is not None and num_items_in_batch > 0:
+        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
+        loss = loss / num_items_in_batch
+    else:
+        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100)
+
+    return loss
```
Comment (muellerzr) [inline src/transformers/loss/loss_utils.py:45]: Everthing that uses `cross_entropy` will need this fix. So let's change this up some. 

I think we should define our own `CrossEntropyLoss` that we call, that does the following:

```python
def CrossEntropyLoss(logits, labels, num_items_in_batch=None, **kwargs):
    reduction = "sum" if num_items_in_batch is not None else "mean"
    loss = nn.functional.cross_entropy(logits, labels, reduction=reduction, **kwargs)
    if reduction == "sum":
        loss /= num_items_in_batch
````    

We can then call it everywhere we need to

**Comment by Hannibal046** [comment]
Quick question: Is Distributed Data Parallel (DDP) training affected by this issue in the same way?

**Comment by GirinMan** [comment]
@Hannibal046 If you used gradient accumulation with DDP, yes.

**Comment by Hannibal046** [comment]
Consider the following scenario with two samples of different lengths (wo gradient acc):

1. Sample A: 10 tokens
2. Sample B: 5 tokens

- Case 1: Single GPU (batch_size=2)

```python
loss = average_loss(10 tokens + 5 tokens)
gradient_1 = backprop(loss)  # Based on average of 15 tokens
```
- Case 2: Distributed Training (2 GPUs, local_batch_size=1)
```python
# GPU 1
loss_1 = average_loss(10 tokens)
gradient_1 = backprop(loss_1)

# GPU 2
loss_2 = average_loss(5 tokens)
gradient_2 = backprop(loss_2)

final_gradient = 0.5 * gradient_1 + 0.5 * gradient_2
```
so `gradient_1 != final_gradient`.

**Comment by GirinMan** [comment]
The issue fixed in this PR is about gradient accumulation, not DDP.
What you explained about is orthogonal with gradient accumulation, and may be intended behavior.

**Comment by paulcx** [comment]
After using this fix, the following problem occurred:
the grad norm and loss (custom loss from trl cpo trainer) increased dramatically

image
green：llama34b-fullparams-bs4-grad8-wo_fix
orange: llama34b-fullparams-bs4-grad8-wo_fix

**Comment by JianbangZ** [comment]
> After using this fix, the following problem occurred: the grad norm and loss (custom loss from trl cpo trainer) increased dramatically
> 
> image green：llama34b-deepspeed-zero3-8gpus-fullparams-bs4-grad8-wo_fix orange: llama34b-deepspeed-zero3-8gpus-fullparams-bs4-grad8-wo_fix

Can confirm the same issue

**Comment by aliencaocao** [comment]
I think the fix isnt complete, another fix is needed to make it normal. See https://github.com/hiyouga/LLaMA-Factory/issues/5747#issuecomment-2425488226

**Comment by paulcx** [comment]
> I think the fix isnt complete, another fix is needed to make it normal. See [hiyouga/LLaMA-Factory#5747 (comment)](https://github.com/hiyouga/LLaMA-Factory/issues/5747#issuecomment-2425488226)

Hi @ArthurZucker , would you mind check the further fix?

**Comment by paulcx** [comment]
As [hiyouga/LLaMA-Factory#5747 (comment)](https://github.com/hiyouga/LLaMA-Factory/issues/5747#issuecomment-2425488226) suggested, maybe a few changes will make this fix work again?

**Comment by iridescentee** [comment]
> > After using this fix, the following problem occurred: the grad norm and loss (custom loss from trl cpo trainer) increased dramatically
> > image green：llama34b-deepspeed-zero3-8gpus-fullparams-bs4-grad8-wo_fix orange: llama34b-deepspeed-zero3-8gpus-fullparams-bs4-grad8-wo_fix
> 
> Can confirm the same issue

+ 1

**Comment by iridescentee** [comment]
> > After using this fix, the following problem occurred: the grad norm and loss (custom loss from trl cpo trainer) increased dramatically
> > image green：llama34b-deepspeed-zero3-8gpus-fullparams-bs4-grad8-wo_fix orange: llama34b-deepspeed-zero3-8gpus-fullparams-bs4-grad8-wo_fix
> 
> Can confirm the same issue
same. encountered when fine-tuning using qwen2.5.

**Comment by muellerzr** [comment]
Fix is being worked on here: https://github.com/huggingface/transformers/pull/34283

**Comment by thusinh1969** [comment]
Oh my....

**Comment by JaheimLee** [comment]
Hi, I have a question. I think `num_items_in_batch` can be calculated by `torch.where(target != ignore_index, torch.ones_like(target), torch.zeros_like(target)).sum()`, so why do we need to pass it manually? The `reduce` can always be `'sum'`. @muellerzr @ArthurZucker

**Comment by taehyunzzz** [comment]
Hello all, I could not follow the whole context of the PR here. I am guessing a fix is underway regarding the loss becoming huge due to [this code](https://github.com/huggingface/transformers/blob/1d063793318b20654ebb850f48f43e0a247ab7bb/src/transformers/trainer.py#L3605). The code scales the loss by gradient accumulation step in the latest trainer (not in previous version). The snippet is included in a single training iteration step, for which I think scaling the loss by grad_acc_step is not appropriate, if the loss is not de-scaled by grad_acc_step inside the accelerator.backward call in the following line. Is my understanding correct?

**Comment by muellerzr** [comment]
Yes accelerate de-scales the loss when calling `backward()` which is why we do so. (Accelerate has always done this)

**Comment by taehyunzzz** [comment]
@muellerzr Thank you for the reply. I have just one more question to clear things up.

You've mentioned that accelerate had always internally de-scaled the loss by gradient accumulation steps.
However, the previous trainer.py implementations do not scale (scale up) the loss by grad_acc_steps before accelerator.backward. 
If accelerator.backward had been internally de-scaling the loss by grad_acc_steps, does that mean the previous trainer had been training a model on 1/grad_acc_steps * sum(loss) instead of sum(loss) when grad_acc_steps > 1?

File: src/transformers/models/mistral/modeling_mistral.py [hunk 8]
```diff
@@ -1291,27 +1291,8 @@ def forward(
 
         loss = None
         if labels is not None:
-            labels = labels.to(logits.device)
```
Comment (cphillippi-stripe) [inline src/transformers/models/mistral/modeling_mistral.py:1294]: I'm seeing issues w/ multi-gpu sequence classification training after this change (using Mistral). I believe it is due to the removal of this line (which I'm having a very hard time monkeypatching via `loss_function` for some reason). I also see this line very often in quite a few different models here. Were multi-gpu setups tested here for training in sequence classification mode? I'm really curious how this passed if so, because I don't see any lines like this in the new `ForSequenceClassificationLoss` function that `self.loss_function` now seems to resolve to.

I'm seeing tracebacks like this (edited for privacy):
```
...
File ".../training_workflow.py", line 426, in train
216     trainer.train()
217       File ".../transformers/transformers/transformers/trainer.py", line 2123, in train
218     return inner_training_loop(
219            ^^^^^^^^^^^^^^^^^^^^
220       File ".../transformers/transformers/transformers/trainer.py", line 2481, in _inner_training_loop
221     tr_loss_step = self.training_step(model, inputs, num_items_in_batch)
222                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
223       File ".../transformers/transformers/transformers/trainer.py", line 3579, in training_step
224     loss = self.compute_loss(model, inputs, num_items_in_batch=num_items_in_batch)
225            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
226       File ".../transformers/transformers/transformers/trainer.py", line 3633, in compute_loss
227     outputs = model(**inputs)
228               ^^^^^^^^^^^^^^^
229       File ".../torch/torch/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
230     return self._call_impl(*args, **kwargs)
231            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
232       File ".../torch/torch/torch/nn/modules/module.py", line 1541, in _call_impl
233     return forward_call(*args, **kwargs)
234            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
235       File ".../accelerate/accelerate/accelerate/utils/operations.py", line 823, in forward
236     return model_forward(*args, **kwargs)
237            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
238       File ".../accelerate/accelerate/accelerate/utils/operations.py", line 811, in __call__
239     return convert_to_fp32(self.model_forward(*args, **kwargs))
240                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
241       File ".../torch/torch/torch/amp/autocast_mode.py", line 16, in decorate_autocast
242     return func(*args, **kwargs)
243            ^^^^^^^^^^^^^^^^^^^^^
244       File ".../peft/peft/peft/peft_model.py", line 1446, in forward
245     return self.base_model(
246            ^^^^^^^^^^^^^^^^
247       File ".../torch/torch/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
248     return self._call_impl(*args, **kwargs)
249            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
250       File ".../torch/torch/torch/nn/modules/module.py", line 1541, in _call_impl
251     return forward_call(*args, **kwargs)
252            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
253       File ".../peft/peft/peft/tuners/tuners_utils.py", line 197, in forward
254     return self.model.forward(*args, **kwargs)
255            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
256       File ".../accelerate/accelerate/accelerate/hooks.py", line 170, in new_forward
257     output = module._old_forward(*args, **kwargs)
258              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
259       File ".../transformers/transformers/transformers/models/mistral/modeling_mistral.py", line 1200, in forward
260     loss = self.loss_function(logits=logits, labels=labels, pooled_logits=pooled_logits, config=self.config)
261            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
262       File ".../transformers/transformers/transformers/loss/loss_utils.py", line 67, in ForSequenceClassificationLoss
263     loss = fixed_cross_entropy(pooled_logits.view(-1, num_labels), labels.view(-1), **kwargs)
264            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
265       File ".../transformers/transformers/transformers/loss/loss_utils.py", line 26, in fixed_cross_entropy
266     loss = nn.functional.cross_entropy(source, target, ignore_index=ignore_index, reduction=reduction)
267            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
268       File ".../torch/torch/torch/nn/functional.py", line 3086, in cross_entropy
269     return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
```

For now, if you think you have a good way I can just patch this, I'd appreciate it, but replacing `PretrainedModel.loss_function` doesn't seem to do the trick...

**Comment by bauwenst** [comment]
FYI: As also noted by @cphillippi-stripe, the `@property` `PreTrainedModel.loss_function` that was added by this PR breaks all software packages that have models which (1) inherit from `PreTrainedModel` while (2) setting a `loss_function` beforehand. 

I maintain such a package, and indeed, training is now broken and my IDE complains that `loss_function` cannot be set. 

image

The default behaviour of setting the loss to *causal LM* loss when no `loss_type` is defined in the config seems **highly** unwarranted to me. It is assumed now that either models define this brand new `loss_type` field in the config and that otherwise the loss type can be inferred from the `ForXYZ` suffix:

https://github.com/huggingface/transformers/blob/63766abe362b930522dd073c9173499ba5fde02a/src/transformers/modeling_utils.py#L5110-L5120

Yet, this fails to take into account the following cases where `transformers` is used as a dependency:
1. Model classes defined for *tasks that don't have a known auto class* in `transformers`. For example: if your package defines a `ModelForDependencyParsing`, the above regex search will not find its loss and default to causal loss, which is wrong.
 
```python
  File ".../ArchIt/src/archit/instantiation/tasks.py", line 328, in computeLoss
    return self.loss_function(arc_scores, arc_labels) + self.loss_function(rel_scores, rel_labels)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: ForCausalLMLoss() missing 1 required positional argument: 'vocab_size'
```

2. Model classes that have a known task, but it's not in the name (e.g. `ModelForSentenceClassifying`). Unless the config is from pre-December 2024, the config does not define its loss type.

3. Loss is not inherent to a `PreTrainedModel`. There are plenty of papers that start from the model weights of one model and then change its loss function for continued pretraining (take [this one](https://aclanthology.org/2024.acl-long.136/) as a recent example).

TL;DR: Seems like a design flaw to make `loss_function` a `@property` and seems like a mistake to overwrite it to `ForCausalLMLoss` for all models that don't fit the very narrow range of tasks supported by `transformers`.

**Comment by cphillippi-stripe** [comment]
> FYI: As also noted by @cphillippi-stripe, the `@property` `PreTrainedModel.loss_function` that was added by this PR breaks all software packages that have models which (1) inherit from `PreTrainedModel` while (2) setting a `loss_function` beforehand.
> 
> I maintain [such a package](https://github.com/bauwenst/ArchIt/blob/4018ead262e4a6551bb6f1ff21bf6f52af6fce70/src/archit/instantiation/abstracts.py#L234-L236), and indeed, training is now broken and my IDE complains that `loss_function` cannot be set.
> 
> image
> 
> The default behaviour of setting the loss to _causal LM_ loss when no `loss_type` is defined in the config seems **highly** unwarranted to me. It is assumed now that either models define this brand new `loss_type` field in the config and that otherwise the loss type can be inferred from the `ForXYZ` suffix:
> 
> https://github.com/huggingface/transformers/blob/63766abe362b930522dd073c9173499ba5fde02a/src/transformers/modeling_utils.py#L5110-L5120
> 
> Yet, this fails to take into account the following cases where `transformers` is used as a dependency:
> 
> 1. Model classes defined for _tasks that don't have a known auto class_ in `transformers`. For example: if your package defines a `ModelForDependencyParsing`, the above regex search will not find its loss and default to causal loss, which is wrong.
> 
> ```python
>   File ".../ArchIt/src/archit/instantiation/tasks.py", line 328, in computeLoss
>     return self.loss_function(arc_scores, arc_labels) + self.loss_function(rel_scores, rel_labels)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> TypeError: ForCausalLMLoss() missing 1 required positional argument: 'vocab_size'
> ```
> 
> 2. Model classes that have a known task, but it's not in the name (e.g. `ModelForSentenceClassifying`). Unless the config is from pre-October 2024, the config does not define its loss type.
> 3. Loss is not inherent to a `PreTrainedModel`. There are plenty of papers that start from the model weights of one model and then change its loss function for continued pretraining (take [this one](https://aclanthology.org/2024.acl-long.136/) as a recent example).
> 
> TL;DR: Seems like a design flaw to make `loss_function` a `@property` and seems like a mistake to overwrite it to `ForCausalLMLoss` for all models that don't fit the very narrow range of tasks supported by `transformers`.

Not a fix but a temporary workaround for me was this patch after loading a mdoel:
```
model = AutoModelForSequenceClassification.from_pretrained(...)

old_loss_function = model.loss_function
@lru_cache
def get_loss_function(self):
    def fixed_loss(labels, pooled_logits, config, **kwargs):
        labels = labels.to(pooled_logits.device)
        return old_loss_function(labels, pooled_logits, config, **kwargs)
    return fixed_loss
    
model.__class__.loss_function = property(get_loss_function)
```

Also, @ArthurZucker, I believe this fix confirms removing the `labels = labels.to(pooled_logits.device)` line breaks the multi-gpu setup. I would be very surprised if only Mistral is affected here given similar lines were removed from other models. Initially, I tried patched this at the `PreTrainedModel.loss_function` level, but for whatever reason the loaded models don't seem to pick up the change, and I'm not sure why.

**Comment by ArthurZucker** [comment]
Feedback taken, will open a PR to get to something better. It's hard to take everything into account, and support important to get your feedbacks! 🤗 
Will merge the fix to loss on multi-GPU (I don't think we test all tasks for training and that is a flaw in our test environnement for sure! ) cc @ydshieh as for a TODO 😉
